### PR TITLE
Observability : proxy metric

### PR DESCRIPTION
**Pull request by zirain**
xhref: https://github.com/envoyproxy/gateway/issues/699

### CODE REVIEW

File: docs/latest/design/metrics.md [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+- Enable prometheus metric
+- Push metrics via Open Telemetry Sink
+- Customize histogram buckets of target metric
+
+### ProxyMetric API Type
+
+```golang mdox-exec="sed '1,7d' api/config/v1alpha1/metric_types.go"
+type ProxyMetric struct {
+	// Prometheus defines the configuration for Admin endpoint `/stats/prometheus`.
+	Prometheus *PrometheusProvider `json:"prometheus,omitempty"`
```
Comment (arkodg) [inline docs/latest/design/metrics.md:33]: should we create a `Push` / `Pull` based hierarchy here ?
```
telemetry:
  metrics:
    pull:
      type: Prometheus
      prometheus: 
```

```
telemetry:
  metrics:
    push:
      type: OpenTelemetry
      openTelemetry: 
```

File: docs/latest/design/metrics.md [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+	Type MetricSinkType `json:"type"`
+	// OpenTelemetry defines the configuration for OpenTelemetry sink.
+	// It's required if the sink type is OpenTelemetry.
+	OpenTelemetry *OpenTelemetrySink `json:"openTelemetry,omitempty"`
+}
+
+type OpenTelemetrySink struct {
+	// Backend defines the backend to send OpenTelemetry metrics to.
+	// +kubebuilder:default={port: 3417}
+	Backend BackendService `json:"backend"`
```
Comment (arkodg) [inline docs/latest/design/metrics.md:61]: other Telemetry Sink APIs dont seem to have the `Backend` field
https://github.com/envoyproxy/gateway/blob/609949d6dfcc3726a78a7d6cbb7b6298883159fa/api/config/v1alpha1/accesslogging_types.go#L89
https://github.com/envoyproxy/gateway/blob/609949d6dfcc3726a78a7d6cbb7b6298883159fa/api/config/v1alpha1/tracing_types.go#L38
would be good to add the same design philosophy to all of them

File: api/config/v1alpha1/metric_types.go [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+	// +kubebuilder:default={port: 3417}
+	Backend BackendService `json:"backend"`
+
+	// TODO: add support for customizing OpenTelemetry sink in https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/stat_sinks/open_telemetry/v3/open_telemetry.proto#envoy-v3-api-msg-extensions-stat-sinks-open-telemetry-v3-sinkconfig
+}
+
+type PrometheusProvider struct {
+	// Enable defines whether to enable the Prometheus endpoint.
+	// Prometheus' annotations will be added to pod if enabled.
+	Enable bool `json:"enable,omitempty"`
```
Comment (arkodg) [inline api/config/v1alpha1/metric_types.go:46]: do we need an additional `bool` here, or can the instantiation of `PrometheusProvider{}` be adequate enough to confirm intent to enable the prometheus endpoint ?

File: api/config/v1alpha1/metric_types.go [hunk 4]
```diff
@@ -0,0 +1,60 @@
+// Copyright Envoy Gateway Authors
+// SPDX-License-Identifier: Apache-2.0
+// The full text of the Apache license is available in the LICENSE file at
+// the root of the repo.
+
+package v1alpha1
+
+type ProxyMetric struct {
```
Comment (arkodg) [inline api/config/v1alpha1/metric_types.go:8]:
```suggestion
type ProxyMetrics struct {
```

File: api/config/v1alpha1/metric_types.go [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+	// Prometheus defines the configuration for Admin endpoint `/stats/prometheus`.
+	Prometheus *PrometheusProvider `json:"prometheus,omitempty"`
```
Comment (arkodg) [inline api/config/v1alpha1/metric_types.go:10]: should we create a `Push` / `Pull` based hierarchy here ?
```
telemetry:
  metrics:
    type: Pull
    pull:
      type: Prometheus
      prometheus: 
```

```
telemetry:
  metrics:
    type: Push
    push:
      type: OpenTelemetry
      openTelemetry: 
```

File: api/config/v1alpha1/metric_types.go [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+	Type MetricSinkType `json:"type"`
+	// OpenTelemetry defines the configuration for OpenTelemetry sink.
+	// It's required if the sink type is OpenTelemetry.
+	OpenTelemetry *OpenTelemetrySink `json:"openTelemetry,omitempty"`
+}
+
+type OpenTelemetrySink struct {
+	// Backend defines the backend to send OpenTelemetry metrics to.
+	// +kubebuilder:default={port: 3417}
+	Backend BackendService `json:"backend"`
```
Comment (arkodg) [inline api/config/v1alpha1/metric_types.go:38]: other Telemetry Sink APIs dont seem to have the `Backend` field
https://github.com/envoyproxy/gateway/blob/609949d6dfcc3726a78a7d6cbb7b6298883159fa/api/config/v1alpha1/accesslogging_types.go#L89
https://github.com/envoyproxy/gateway/blob/609949d6dfcc3726a78a7d6cbb7b6298883159fa/api/config/v1alpha1/tracing_types.go#L38
would be good to add the same design philosophy to all of them

File: api/config/v1alpha1/metric_types.go (hunk 3 above)
Comment (zirain) [reply api/config/v1alpha1/metric_types.go:46]: that's fine to me, do you think it's fine to allow users custome the path of metric endpoint,
envoy use `/stats/prometheus`, but a lot of app use `/metrics`.

File: api/config/v1alpha1/metric_types.go (hunk 6 above)
Comment (zirain) [reply api/config/v1alpha1/metric_types.go:38]: ok, let's keep same first, and change all of them later.

File: api/config/v1alpha1/metric_types.go (hunk 5 above)
Comment (zirain) [reply api/config/v1alpha1/metric_types.go:10]: Actually, they can exist at the same time.

**Comment by tmsnan** [comment]
https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/metrics/v3/stats.proto#envoy-v3-api-msg-config-metrics-v3-statsmatcher
Whether `stats_matcher`  field should be provided to control the number of stats ?

**Comment by zirain** [comment]
> https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/metrics/v3/stats.proto#envoy-v3-api-msg-config-metrics-v3-statsmatcher Whether `stats_matcher` field should be provided to control the number of stats ?

we're close to v0.5.0-rc1, maybe add them in next version.

File: api/config/v1alpha1/metric_types.go (hunk 5 above)
Comment (arkodg) [reply api/config/v1alpha1/metric_types.go:10]: ah I see

Comment (arkodg) [reply api/config/v1alpha1/metric_types.go:10]: if thats the case do you think this format is handy or too wordy ?
this way we allow the user to specify both push and pull but only one for each type
```
telemetry:
  metrics:
    push:
      type: OpenTelemetry
      openTelemetry: 
    pull:
      type: Prometheus
      prometheus: 
```

File: tools/make/kube.mk [hunk 7]
```diff
@@ -30,7 +30,12 @@ CONTROLLERGEN_OBJECT_FLAGS :=  object:headerFile="$(ROOT_DIR)/tools/boilerplate/
 .PHONY: manifests
 manifests: $(tools/controller-gen) generate-gwapi-manifests ## Generate WebhookConfiguration, ClusterRole and CustomResourceDefinition objects.
 	@$(LOG_TARGET)
-	$(tools/controller-gen) rbac:roleName=envoy-gateway-role crd webhook paths="./..." output:crd:artifacts:config=charts/gateway-helm/crds/generated output:rbac:artifacts:config=charts/gateway-helm/templates/generated/rbac output:webhook:artifacts:config=charts/gateway-helm/templates/generated/webhook
+	$(tools/controller-gen) rbac:roleName=envoy-gateway-role \
+	crd:allowDangerousTypes=true \
```
Comment (arkodg) [inline tools/make/kube.mk:34]: why is `allowDangerousTypes` needed ?

File: api/config/v1alpha1/metric_types.go (hunk 5 above)
Comment (arkodg) [reply api/config/v1alpha1/metric_types.go:10]: @AliceProxy @kflynn and I discussed this in the community meeting today and feel like `pull` and `push` might be redundant keywords to type, resolving this comment

File: docs/latest/design/metrics.md [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+
+```golang mdox-exec="sed '1,7d' api/config/v1alpha1/metric_types.go"
+type ProxyMetrics struct {
+	// Prometheus defines the configuration for Admin endpoint `/stats/prometheus`.
+	Prometheus *PrometheusProvider `json:"prometheus,omitempty"`
+	// Sinks defines the metric sinks where metrics are sent to.
+	Sinks []MetricSink `json:"sinks,omitempty"`
+	// HistogramBucketSettings defines rules for setting the histogram buckets.
+	// Default buckets are used if not set. See more details at https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/metrics/v3/stats.proto.html#config-metrics-v3-histogrambucketsettings.
+	HistogramBucketSettings []HistogramBucketSetting `json:"histogramBucketSettings,omitempty"`
```
Comment (arkodg) [inline docs/latest/design/metrics.md:38]: should we add this field as a follow up ?

Comment (zirain) [reply docs/latest/design/metrics.md:38]: sure, we can put this and `statsMatcher` as follow up, that will make this proposal simpler.

Comment (arkodg) [reply docs/latest/design/metrics.md:38]: +1 to this

Comment (zirain) [reply docs/latest/design/metrics.md]: create a issue for this: https://github.com/envoyproxy/gateway/issues/1669

**Review by arkodg — APPROVED** [review]
LGTM thanks !

**Review by AliceProxy — APPROVED** [review]
lgtm
